File: crates/aquarium/src/platform/windows.rs
use super::Platform;
use crate::window::{Window, WindowParams};
use raw_window_handle::Win32WindowHandle;
use std::num::NonZeroIsize;
use windows::{
    core::{w, HSTRING},
    Win32::{
        Foundation::{HWND, LPARAM, LRESULT, WPARAM},
        Graphics::Gdi::ValidateRect,
        System::LibraryLoader::GetModuleHandleW,
        UI::WindowsAndMessaging::{
            CreateWindowExW, DefWindowProcW, DispatchMessageW, GetMessageW, LoadCursorW,
            PostQuitMessage, RegisterClassW, ShowWindow, CS_HREDRAW, CS_VREDRAW, CW_USEDEFAULT,
            IDC_ARROW, MSG, SW_SHOW, WINDOW_EX_STYLE, WM_DESTROY, WM_PAINT, WNDCLASSW,
            WS_OVERLAPPEDWINDOW, WS_VISIBLE,
        },
    },
};

pub struct WindowsPlatform {}

impl WindowsPlatform {
    pub fn new() -> Self {
        Self {}
    }
}

impl Default for WindowsPlatform {
    fn default() -> Self {
        WindowsPlatform::new()
    }
}

impl Platform for WindowsPlatform {
    fn new_window(&self, params: WindowParams) -> Window {
        unsafe {
            let instance = GetModuleHandleW(None).unwrap();
            assert!(instance.0 != 0);

            let window_class = w!("Aquarium::Window");
            let wc = WNDCLASSW {
                hCursor: LoadCursorW(None, IDC_ARROW).unwrap(),
                hInstance: instance.into(),
                lpszClassName: window_class,

                style: CS_HREDRAW | CS_VREDRAW,
                lpfnWndProc: Some(wndproc),
                ..Default::default()
            };

            let atom = RegisterClassW(&wc);
            assert!(atom != 0);

            // TODO: Eventually make more things than title configurable lol
            let raw_hwnd = CreateWindowExW(
                WINDOW_EX_STYLE::default(),
                window_class,
                &HSTRING::from(params.title),
                WS_OVERLAPPEDWINDOW | WS_VISIBLE,
                CW_USEDEFAULT,
                CW_USEDEFAULT,
                params.width.into(),
                params.height.into(),
                None,
                None,
                instance,
                None,
            );

            ShowWindow(raw_hwnd, SW_SHOW).unwrap();

            let window_handle = Win32WindowHandle::new(NonZeroIsize::new(raw_hwnd.0).unwrap());

            Window {}
        }
    }

    fn start(&self, on_start: Box<dyn 'static + FnOnce()>) {
        on_start();

        let mut message = MSG::default();
        unsafe {
            while GetMessageW(&mut message, None, 0, 0).into() {
                DispatchMessageW(&message);
            }
        }
    }
}

extern "system" fn wndproc(window: HWND, message: u32, wparam: WPARAM, lparam: LPARAM) -> LRESULT {
    unsafe {
        match message {
            WM_PAINT => {
                ValidateRect(window, None).unwrap();
                LRESULT(0)
            }
            WM_DESTROY => {
                PostQuitMessage(0);
                LRESULT(0)
            }
            _ => DefWindowProcW(window, message, wparam, lparam),
        }
    }
}

File: src/engine/windowing/platforms/windows.zig
const std = @import("std");
const win32 = @import("win32").everything;

pub const WindowManager = struct {
    hInstance: win32.HINSTANCE,

    pub fn init() WindowManager {
        const hInstance = win32.GetModuleHandleW(null).?;
        return WindowManager{ .hInstance = hInstance };
    }

    pub fn run(self: *WindowManager) void {
        _ = self;
        var msg: win32.MSG = undefined;
        while (win32.GetMessageW(&msg, null, 0, 0) != 0) {
            _ = win32.TranslateMessage(&msg);
            _ = win32.DispatchMessageW(&msg);
        }
    }

    pub fn deinit(self: *WindowManager) void {
        self.* = undefined;
    }
};

pub const Window = struct {
    hInstance: win32.HINSTANCE,
    hwnd: win32.HWND,

    pub fn init(window_manager: *WindowManager) Window {
        const CLASS_NAME = win32.L("Manatee");
        const wc = win32.WNDCLASSEXW{
            .cbSize = @sizeOf(win32.WNDCLASSEXW),
            .style = win32.WNDCLASS_STYLES{},
            .lpfnWndProc = WindowProc,
            .cbClsExtra = 0,
            .cbWndExtra = 0,
            .hInstance = @ptrCast(window_manager.hInstance),
            .hIcon = null,
            .hCursor = null,
            .hbrBackground = null,
            .lpszMenuName = null,
            .lpszClassName = CLASS_NAME,
            .hIconSm = null,
        };

        _ = win32.RegisterClassExW(&wc);

        // I hate that the Zig standard formatter won't let me add line breaks to func calls lol
        const hwnd = win32.CreateWindowExW(win32.WS_EX_OVERLAPPEDWINDOW, CLASS_NAME, win32.L("Manatee Game Engine Window"), win32.WS_OVERLAPPEDWINDOW, win32.CW_USEDEFAULT, win32.CW_USEDEFAULT, 400, 200, null, null, window_manager.hInstance, null);

        _ = win32.ShowWindow(hwnd, win32.SW_SHOW);

        return Window{
            .hInstance = window_manager.hInstance,
            .hwnd = hwnd.?,
        };
    }

    pub fn deinit(self: *Window) void {
        self.* = undefined;
    }

    fn WindowProc(
        hwnd: win32.HWND,
        uMsg: u32,
        wParam: win32.WPARAM,
        lParam: win32.LPARAM,
    ) callconv(std.os.windows.WINAPI) win32.LRESULT {
        switch (uMsg) {
            win32.WM_DESTROY => {
                win32.PostQuitMessage(0);
                return 0;
            },
            win32.WM_PAINT => {
                return win32.ValidateRect(hwnd, null);
            },
            else => {
                return win32.DefWindowProcW(hwnd, uMsg, wParam, lParam);
            },
        }
    }
};
